File: Data_Structures/Linked_lists/traversal_modification.c
// A program to traverse through the list and double the value of each node

#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the linked list
void insertAtEnd(struct Node** head, int data) {
    struct Node* newNode = createNode(data);
    if (*head == NULL) {
        *head = newNode;
        return;
    }
    struct Node* last = *head;
    while (last->next != NULL) {
        last = last->next;
    }
    last->next = newNode;
}

// Function to display the linked list
void displayList(struct Node* head) {
    if (head == NULL) {
        printf("Linked list is empty.\n");
        return;
    }
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

// Function to traverse and modify the linked list by doubling the value of each node
void doubleList(struct Node* head) {
    struct Node* current = head;
    while (current != NULL) {
        current->data *= 2;
        current = current->next;
    }
}

int main() {
    struct Node* head = NULL;
    insertAtEnd(&head, 1);
    insertAtEnd(&head, 2);
    insertAtEnd(&head, 3);
    insertAtEnd(&head, 4);
    insertAtEnd(&head, 5);

    printf("Original Linked List: ");
    displayList(head);

    doubleList(head);

    printf("Modified Linked List (doubled values): ");
    displayList(head);

    printf("\n");
    return 0;
}

File: Data_Structures/Linked_lists/traversal_modification.py
# A program to traverse through the list and double the value of each node

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Function to insert a node at the end of the linked list
def insertAtEnd(head, data):
    newNode = Node(data)
    if head is None:
        head = newNode
        return head
    last = head
    while last.next is not None:
        last = last.next
    last.next = newNode
    return head

# Function to display the linked list
def displayList(head):
    if head is None:
        print("Linked list is empty.")
        return
    temp = head
    while temp is not None:
        print(temp.data, end=" ")
        temp = temp.next
    print()

# Function to traverse and modify the linked list by doubling the value of each node
def doubleList(head):
    current = head
    while current is not None:
        current.data *= 2
        current = current.next

# Main function
def main():
    head = None
    head = insertAtEnd(head, 1)
    head = insertAtEnd(head, 2)
    head = insertAtEnd(head, 3)
    head = insertAtEnd(head, 4)
    head = insertAtEnd(head, 5)

    print("Original Linked List:", end=" ")
    displayList(head)

    doubleList(head)

    print("Modified Linked List (doubled values):", end=" ")
    displayList(head)

if __name__ == '__main__':
    main()
